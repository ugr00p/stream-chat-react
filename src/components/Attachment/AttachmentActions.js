import React from 'react';
import PropTypes from 'prop-types';

/**
 * AttachmentActions - The actions you can take on an attachment
 *
 * @example ../../docs/AttachmentActions.md
 * @type {React.FC<import('type').AttachmentActionsProps>}
 */
const AttachmentActions = ({ text, id, actions, actionHandler }) => (
  <div className="str-chat__message-attachment-actions">
    <div className="str-chat__message-attachment-actions-form">
      <span key={0}>{text}</span>
      {actions.map((action) => (
        <button
          className={`str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--${action.style}`}
          data-testid={`${action.name}`}
          key={`${id}-${action.value}`}
          data-value={action.value}
          onClick={(e) => actionHandler(action.name, action.value, e)}
        >
          {action.text}
        </button>
      ))}
    </div>
  </div>
);

AttachmentActions.propTypes = {
  /** Unique id for action button key. Key is generated by concatenating this id with action value - {`${id}-${action.value}`} */
  id: PropTypes.string.isRequired,
  /** The text for the form input */
  text: PropTypes.string,
  /** A list of actions */
  actions: PropTypes.array.isRequired,
  /**
   *
   * Handler for actions. Actions in combination with attachments can be used to build [commands](https://getstream.io/chat/docs/#channel_commands).
   *
   * @param name {string} Name of action
   * @param value {string} Value of action
   * @param event Dom event that triggered this handler
   */
  actionHandler: PropTypes.func.isRequired,
};

export default React.memo(AttachmentActions);
