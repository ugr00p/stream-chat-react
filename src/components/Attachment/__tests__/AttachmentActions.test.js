import React from 'react';
import '@testing-library/jest-dom';
import { render, waitFor, fireEvent } from '@testing-library/react';

import renderer from 'react-test-renderer';
import { v4 as uuidv4 } from 'uuid';
import AttachmentActions from '../AttachmentActions';

const getComponent = (props) => {
  return <AttachmentActions {...props} />;
};
const actions = [
  {
    value: 'action 1',
    name: 'action 1 name',
    text: 'action 1 text',
  },
  {
    value: 'action 2',
    name: 'action 2 name',
    text: 'action 2 text',
  },
];

describe('AttachmentActions', () => {
  it('should render AttachmentActions component', () => {
    const tree = renderer
      .create(
        getComponent({
          actions,
          id: uuidv4(),
          actionHandler: jest.fn(),
        }),
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should call actionHandler on click', async () => {
    const actionHandler = jest.fn();
    const { getByTestId } = render(
      getComponent({
        actions,
        id: uuidv4(),
        actionHandler,
      }),
    );

    await waitFor(() => {
      expect(getByTestId(actions[0].name)).toBeInTheDocument();
    });

    fireEvent.click(getByTestId(actions[0].name));
    fireEvent.click(getByTestId(actions[1].name));

    await waitFor(() => {
      // eslint-disable-next-line jest/prefer-called-with
      expect(actionHandler).toHaveBeenCalledTimes(2);
    });
  });
});
