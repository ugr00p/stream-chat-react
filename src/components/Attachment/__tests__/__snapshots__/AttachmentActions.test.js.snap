// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentActions should render AttachmentActions component 1`] = `
<div
  className="str-chat__message-attachment-actions"
>
  <div
    className="str-chat__message-attachment-actions-form"
  >
    <span />
    <button
      className="str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--undefined"
      data-testid="action 1 name"
      data-value="action 1"
      onClick={[Function]}
    >
      action 1 text
    </button>
    <button
      className="str-chat__message-attachment-actions-button str-chat__message-attachment-actions-button--undefined"
      data-testid="action 2 name"
      data-value="action 2"
      onClick={[Function]}
    >
      action 2 text
    </button>
  </div>
</div>
`;
